Сделана часть задачи - русский, казахский, Declension

kruprog opened the pull request:
Для русского реализованы все юнит тесты Declension которые присутствовали в проекте на c#.
Для казахского реализован пока только один тест.
Сконфигурирован github action для запуска тестов. Запуск вручную по клику.
Сконфигурирован composer для публикации пакета.

kruprog committed:
created composer config and folders

kruprog committed:
rus and qazaq

russian Declension - api and unit tests  completely done.
quazaq Declension - api and unit tests partially done.

kruprog committed:
Update .gitignore

kruprog committed:
added github actions for tests

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
composer to action

kruprog committed:
composer install

kruprog committed:
file not found fix

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
Update DeclensionRussianTest.php

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
simplified

kruprog committed:
composer cache experiment

https://github.com/actions/cache/blob/main/examples.md#php---composer

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
beauty

bzaar commented on src/russian/Gender.php:
```diff
@@ -0,0 +1,23 @@
+    const Neuter='Neuter';
+    const Plural='Plural';
+
+    const decode_array=['Мужской'=>self::Masculine,'Женский'=>self::Feminine, 'Средний'=>self::Neuter, 'Множественное'=>self::Plural];   
```
Множественное число передается при помощи пустого тега: https://morpher.ru/ws3/#rod

kruprog replied:
const decode_array=['Мужской'=>self::Masculine,'Женский'=>self::Feminine, 'Средний'=>self::Neuter, ''=>self::Plural];  
так надо? Так эта конструкция работать не будет, ключ не может быть пустой. По другому напишу.

bzaar commented on tests/DeclensionRussianTest.php:
```diff
@@ -0,0 +1,235 @@
+
+
+
+    public function testParse_Success(): void
```
Не очень понял, чем этот тест отличается от предыдущего (testParse_SuccessMockery)?

kruprog replied:
Верно замечено, один надо убрать. Недосмотрел.

bzaar commented on tests/DeclensionRussianTest.php:
```diff
@@ -0,0 +1,235 @@
+//use Morpher\Ws3Client\Morpher;
+use Morpher\Ws3Client\Russian as Russian;
+
+final class DeclensionRussianTest extends TestCase
```
Прошу переименовать класс в RussianDeclensionTest, а еще лучше - сделать внутри tests три подкаталога (russian, ukrainian, qazaq), как в src, и тогда название класса можно сократить до DeclensionTest.

kruprog replied:
>RussianDeclensionTest
- можно

>сделать внутри tests три подкаталога
- можно попробовать, если это переварит PHPUnit

kruprog replied:
Подкаталоги сделал, переименовать класс в RussianDeclensionTest - сделал,
но сокращать до DeclensionTest  - проблематично, получается несколько классов с одинаковым именем, или тогда указывать разные namespace, в общем ненужное усложнение.

bzaar commented on tests/DeclensionRussianTest.php:
```diff
@@ -0,0 +1,235 @@
+        $lemma='тест';
+
+        $WebClientMock=Mockery::mock(Morpher\Ws3Client\WebClientBase::class);
+        //$WebClientMock->shouldReceive('get_request')->with("/russian/declension",['s'=>$lemma])->andReturn($return_text);
```
Нужно тестировать не только возвращаемые методом Parse данные, но и проверять, что именно Parse передает объекту WebClientMock, в частности:
1. что токен помещен в нужный заголовок (Authorization: Basic) и закодирован при помощи BASE64
2. что при сложении base URL (https://ws3.morpher.ru/) и, например, /russian/declension, не получилось подряд 2 слеша (//)
3. что переданы нужные параметры (?s=кошка) (я вижу, здесь была попытка это проверить, но почему-то не доведена до конца)
4. что переданы нужные заголовки (Accept: application/json)

С этой целью предлагаю вместо get_request сделать метод send(httpRequest), где request - объект класса HttpRequest:

class HttpRequest
{
    public readonly ?string $Url;
    public readonly ?string $Method; // GET / POST / DELETE...
    public readonly ?array $Headers;
    public readonly ?array $QueryParameters;
}

kruprog replied:
>что переданы нужные параметры (?s=кошка) (я вижу, здесь была попытка это проверить, но почему-то не доведена до конца)
причина в том что я хотел обговорить как это требуется сделать.
В основном я основывался на тестах которые были написаны на c#, и там таких проверок я не увидел.
Сейчас как я понял мы идём к тому что тесты на PHP будут сложнее чем были на C#.

>С этой целью предлагаю вместо get_request сделать метод send(httpRequest), где request - объект класса HttpRequest:
Хорошо, так сделаем.
Будем отдельно устанавливать все параметры запроса и проверять их,
и отдельно передавать заполненную структуру на исполнение запроса.

bzaar commented on src/WebClient.php:
```diff
@@ -0,0 +1,61 @@
+		curl_setopt($ch,
+					CURLOPT_HTTPHEADER,
+					array('Accept: application/json',
+						  'Authorization: Basic '.$this->_token));
```
Закодировать токен при помощи BASE64: https://morpher.ru/ws3/#authentication

kruprog replied:
Я ещё подумываю чтобы использовать библиотеку такую как guzzle, для работы с запросами.
Пока просто использовал код из вашего примера, почти ничего там не менял.

bzaar replied:
А в чем преимущество Guzzle перед curl?

kruprog replied:
Лучше почитать про Guzzle.
Guzzle это обёртка, которая внутри может использовать и curl и другие механизмы для работы с http.
Вообще говоря, ей curl не требуется.
Чем лучше?
Проще и короче синтаксис, больше полезных функций, есть встроенный mock для тестирования запросов.

bzaar reviewed (changes requested):
Михаил, спасибо за PR! См. мои комментарии.

kruprog committed:
множественное пустым тегом

kruprog committed:
тест удалил дубликат

kruprog committed:
разделил тесты по папкам

kruprog committed:
Переделаны http запросы

Применена библиотека GuzzleHttp.
Переделана функция вызова запроса.
Тесты дополнены проверками параметров запроса.

kruprog committed:
Тесты переписаны с использованием Guzzle Mock.

Убран mockery.
Убраны ненужные webClientBase и HttpRequest.

kruprog committed:
Казахский тест доделан
